transactionoutbox-core/src/main/java/com/gruelbox/transactionoutbox/TransactionOutboxImpl.java

```diff
@@ -111,23 +114,34 @@ public ParameterizedScheduleBuilder with() {
   @SuppressWarnings("UnusedReturnValue")
   @Override
   public boolean flush() {
+    return flush(false);
+  }
+
+  @Override
+  public boolean flushOrdered() {
+    return flush(true);
+  }
+
+  private boolean flush(boolean isOrdered) {
     if (!initialized.get()) {
       throw new IllegalStateException("Not initialized");
     }
     Instant now = clockProvider.getClock().instant();
-    List<TransactionOutboxEntry> batch = flush(now);
+    List<TransactionOutboxEntry> batch = flush(now, isOrdered);
     expireIdempotencyProtection(now);
     return !batch.isEmpty();
   }
 
-  private List<TransactionOutboxEntry> flush(Instant now) {
+  private List<TransactionOutboxEntry> flush(Instant now, boolean isOrdered) {
     log.debug("Flushing stale tasks");
     var batch =
         transactionManager.inTransactionReturns(
             transaction -> {
               List<TransactionOutboxEntry> result = new ArrayList<>(flushBatchSize);
-              uncheckedly(() -> persistor.selectBatch(transaction, flushBatchSize, now))
-                  .forEach(
+              uncheckedly(() -> isOrdered ?
```

Comment:
You're probably going to need a standalone single-thread `ExecutorService` for this - otherwise the tasks will run in parallel, even if you submit them in order.

Reply:
Didn't get about the concurrency problem tbh, could you elaborate pls?
Here, we have only 1st task per group selected in a single `flush` iteration, and it should impossible to select the 2nd one before completing the 1st because of `processed` condition.
So I thought the problem can occur only in a multi-instance environment which could be solved as it did in the existing (batched) select (with locks etc.)

About future chains - is it for a solution when we select all tasks of each group at one call instead?  I'm going to experiment with this approach.

Reply:
Yes, so the idea is that you select all tasks for all groups, but queue up each group separately, something like this:

```
ConcurrentMap<String, CompletableFuture<Void>> tasks = new ConcurrentMap<>();
...
persistor.selectBatchOrdered(transaction, flushBatchSize, now)
  .forEach(entry ->
    tasks.compute(
      entry.groupId(),
      (groupId, future) -> future == null
        ? CompletableFuture.runAsync(() -> processNow(entry), executor)
        : future.thenRunAsync(() -> processNow(entry), executor));
```

This ignores the presence of `Submitter`, which complicates things slightly, but it demonstrates my thinking/

The result of this is a chain of `CompletableFuture`s for each group, resulting in each task in a group running after the last task in the group completes.